File: services/reservation/src/main/java/com/example/iter/reservation/service/RentalHistoryService.java
package com.example.iter.reservation.service;

import com.example.iter.auth.api.UserQueryPort;
import com.example.iter.auth.api.UserSummary;
import com.example.iter.common.dto.request.PagingRequest;
import com.example.iter.common.dto.response.PageResponse;
import com.example.iter.common.exception.CustomException;
import com.example.iter.common.exception.ErrorCode;
import com.example.iter.device.api.EquipmentInfo;
import com.example.iter.device.api.EquipmentQueryPort;
import com.example.iter.device.api.EquipmentThumbnailQueryPort;
import com.example.iter.reservation.domain.entity.Rental;
import com.example.iter.reservation.domain.repository.RentalHistoryRepository;
import com.example.iter.reservation.domain.repository.spec.RentalSpecifications;
import com.example.iter.reservation.dto.request.RentalHistorySearchRequest;
import com.example.iter.reservation.dto.response.RentalHistoryResponse;
import com.example.iter.reservation.util.RentalHistoryMapper;
import com.example.iter.reservation.util.RentalOverduePolicy;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class RentalHistoryService {

    private final RentalHistoryRepository rentalHistoryRepository;
    private final EquipmentQueryPort equipmentQueryPort;
    private final EquipmentThumbnailQueryPort equipmentThumbnailQueryPort;
    private final UserQueryPort userQueryPort;
    private final RentalHistoryMapper rentalHistoryMapper;

    // 로그인 사용자가 빌린 장비 이력을 조회합니다.
    @Transactional(readOnly = true)
    public PageResponse<RentalHistoryResponse> getBorrowedHistory(Long renterId, RentalHistorySearchRequest request) {
        Page<Rental> rentals = rentalHistoryRepository.findAll(
                RentalSpecifications.borrowedHistory(renterId, request.status(), normalizeKeyword(request.equipmentName())),
                historyPageable(request.page(), request.size())
        );

        return toBorrowedHistoryResponse(rentals, LocalDate.now());
    }

    // 로그인 사용자가 빌려준 장비 이력을 조회합니다.
    @Transactional(readOnly = true)
    public PageResponse<RentalHistoryResponse> getLentHistory(Long ownerId, RentalHistorySearchRequest request) {
        Page<Rental> rentals = rentalHistoryRepository.findLentHistory(
                ownerId,
                request.status(),
                normalizeKeyword(request.equipmentName()),
                historyPageable(request.page(), request.size())
        );

        return toLentHistoryResponse(rentals, LocalDate.now());
    }

    // 로그인 사용자가 빌린 장비 중 현재 연체 중인 거래를 조회합니다.
    @Transactional(readOnly = true)
    public PageResponse<RentalHistoryResponse> getBorrowedOverdueHistory(Long renterId, PagingRequest request) {
        LocalDate today = LocalDate.now();

        Page<Rental> rentals = rentalHistoryRepository.findByRenterIdAndEndDateBeforeAndStatusIn(
                renterId,
                today,
                RentalOverduePolicy.statuses(),
                overduePageable(request.page(), request.size())
        );

        return toBorrowedHistoryResponse(rentals, today);
    }

    // 로그인 사용자가 빌려준 장비 중 현재 연체 중인 거래를 조회합니다.
    @Transactional(readOnly = true)
    public PageResponse<RentalHistoryResponse> getLentOverdueHistory(Long ownerId, PagingRequest request) {
        LocalDate today = LocalDate.now();

        Page<Rental> rentals = rentalHistoryRepository.findByOwnerIdSnapshotAndEndDateBeforeAndStatusIn(
                ownerId,
                today,
                RentalOverduePolicy.statuses(),
                overduePageable(request.page(), request.size())
        );

        return toLentHistoryResponse(rentals, today);
    }

    // ============================================================

    private PageResponse<RentalHistoryResponse> toBorrowedHistoryResponse(Page<Rental> rentals, LocalDate today) {
        if (rentals.isEmpty()) {
            return emptyResponse(rentals);
        }

        Set<Long> equipmentIds = rentals.getContent().stream().map(Rental::getEquipmentId).collect(Collectors.toSet());
        Map<Long, EquipmentInfo> equipmentMap = equipmentQueryPort.findAll(equipmentIds);

        Set<Long> ownerIds = rentals.getContent().stream()
                .map(rental -> getEquipment(equipmentMap, rental.getEquipmentId()).ownerId())
                .collect(Collectors.toSet());

        Map<Long, UserSummary> userMap = userQueryPort.findSummaries(ownerIds);
        Map<Long, String> thumbnailMap = loadThumbnails(equipmentIds);

        List<RentalHistoryResponse> responses = rentals.getContent().stream().map(rental -> {
                EquipmentInfo equipment = getEquipment(equipmentMap, rental.getEquipmentId());
                UserSummary owner = getUser(userMap, equipment.ownerId());

                return rentalHistoryMapper.toResponse(
                        rental,
                        owner,
                        thumbnailMap.get(rental.getEquipmentId()),
                        RentalOverduePolicy.calculateDays(rental, today)
                );
        }).toList();

        return toPageResponse(rentals, responses);
    }

    private PageResponse<RentalHistoryResponse> toLentHistoryResponse(Page<Rental> rentals, LocalDate today) {
        if (rentals.isEmpty()) {
            return emptyResponse(rentals);
        }

        Set<Long> renterIds = rentals.getContent().stream()
                .map(Rental::getRenterId)
                .collect(Collectors.toSet());

        Map<Long, UserSummary> userMap = userQueryPort.findSummaries(renterIds);
        Map<Long, String> thumbnailMap = loadThumbnails(rentals.getContent().stream().map(Rental::getEquipmentId).collect(Collectors.toSet()));

        List<RentalHistoryResponse> responses = rentals.getContent().stream().map(rental -> rentalHistoryMapper.toResponse(
                rental,
                getUser(userMap, rental.getRenterId()),
                thumbnailMap.get(rental.getEquipmentId()),
                RentalOverduePolicy.calculateDays(rental, today)
        )).toList();

        return toPageResponse(rentals, responses);
    }

    private Map<Long, String> loadThumbnails(Collection<Long> equipmentIds) {
        return equipmentThumbnailQueryPort.findThumbnailUrls(equipmentIds);
    }

    private EquipmentInfo getEquipment(Map<Long, EquipmentInfo> equipmentMap, Long equipmentId) {
        EquipmentInfo equipment = equipmentMap.get(equipmentId);

        if (equipment == null) {
            throw new CustomException(ErrorCode.EQUIPMENT_NOT_FOUND);
        }

        return equipment;
    }

    private UserSummary getUser(Map<Long, UserSummary> userMap, Long userId) {
        UserSummary user = userMap.get(userId);

        if (user == null) {
            throw new CustomException(ErrorCode.USER_NOT_FOUND);
        }

        return user;
    }

    private Pageable historyPageable(int page, int size) {
        return PageRequest.of(
                page,
                size,
                Sort.by(Sort.Direction.DESC, "createdAt").and(Sort.by(Sort.Direction.DESC, "id"))
        );
    }

    private Pageable overduePageable(int page, int size) {
        return PageRequest.of(
                page,
                size,
                Sort.by(Sort.Direction.ASC, "endDate")
                        .and(Sort.by(Sort.Direction.DESC, "createdAt"))
                        .and(Sort.by(Sort.Direction.DESC, "id"))
        );
    }

    private String normalizeKeyword(String keyword) {
        if (keyword == null || keyword.isBlank()) {
            return null;
        }

        return keyword.trim();
    }

    private PageResponse<RentalHistoryResponse> emptyResponse(Page<Rental> rentals) {
        return new PageResponse<>(
                List.of(),
                rentals.getNumber(),
                rentals.getSize(),
                rentals.getTotalElements(),
                rentals.getTotalPages()
        );
    }

    private PageResponse<RentalHistoryResponse> toPageResponse(Page<Rental> rentals, List<RentalHistoryResponse> responses) {
        return new PageResponse<>(
                responses,
                rentals.getNumber(),
                rentals.getSize(),
                rentals.getTotalElements(),
                rentals.getTotalPages()
        );
    }
}

File: services/reservation/src/main/kotlin/com/example/iter/reservation/service/RentalHistoryService.kt
package com.example.iter.reservation.service

import com.example.iter.auth.api.UserQueryPort
import com.example.iter.auth.api.UserSummary
import com.example.iter.common.dto.request.PagingRequest
import com.example.iter.common.dto.response.PageResponse
import com.example.iter.common.exception.CustomException
import com.example.iter.common.exception.ErrorCode
import com.example.iter.device.api.EquipmentThumbnailQueryPort
import com.example.iter.reservation.domain.entity.Rental
import com.example.iter.reservation.domain.repository.RentalHistoryRepository
import com.example.iter.reservation.domain.repository.spec.RentalSpecifications
import com.example.iter.reservation.dto.request.RentalHistorySearchRequest
import com.example.iter.reservation.dto.response.RentalHistoryResponse
import com.example.iter.reservation.util.RentalHistoryMapper
import com.example.iter.reservation.util.RentalOverduePolicy
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Clock
import java.time.LocalDate

@Service
class RentalHistoryService(
    private val rentalHistoryRepository: RentalHistoryRepository,
    private val equipmentThumbnailQueryPort: EquipmentThumbnailQueryPort,
    private val userQueryPort: UserQueryPort,
    private val rentalHistoryMapper: RentalHistoryMapper,
    private val clock: Clock
) {

    // 대여자가 빌린 거래를 상태·장비명으로 검색하고 최신 거래부터 반환합니다.
    @Transactional(readOnly = true)
    fun getBorrowedHistory(renterId: Long, request: RentalHistorySearchRequest): PageResponse<RentalHistoryResponse> {
        val rentals = rentalHistoryRepository.findAll(
            RentalSpecifications.borrowedHistory(
                renterId,
                request.status,
                normalizeKeyword(request.equipmentName)
            ),
            historyPageable(request.page, request.size)
        )

        return toBorrowedHistoryResponse(rentals, LocalDate.now(clock))
    }

    // 현재 장비 소유자가 아니라 거래 생성 당시 등록자 스냅샷을 기준으로 빌려준 이력을 조회합니다.
    @Transactional(readOnly = true)
    fun getLentHistory(ownerId: Long, request: RentalHistorySearchRequest): PageResponse<RentalHistoryResponse> {
        val rentals = rentalHistoryRepository.findLentHistory(
            ownerId,
            request.status,
            normalizeKeyword(request.equipmentName),
            historyPageable(request.page, request.size)
        )

        return toLentHistoryResponse(rentals, LocalDate.now(clock))
    }

    // 종료일이 지났고 아직 반납이 끝나지 않은 대여자의 거래를 오래 연체된 순서로 조회합니다.
    @Transactional(readOnly = true)
    fun getBorrowedOverdueHistory(renterId: Long, request: PagingRequest): PageResponse<RentalHistoryResponse> {
        val today = LocalDate.now(clock)
        val rentals = rentalHistoryRepository.findByRenterIdAndEndDateBeforeAndStatusIn(
            renterId,
            today,
            RentalOverduePolicy.statuses(),
            overduePageable(request.page(), request.size())
        )

        return toBorrowedHistoryResponse(rentals, today)
    }

    // 거래 당시 등록자 스냅샷을 기준으로 빌려준 연체 이력을 조회합니다.
    @Transactional(readOnly = true)
    fun getLentOverdueHistory(ownerId: Long, request: PagingRequest): PageResponse<RentalHistoryResponse> {
        val today = LocalDate.now(clock)
        val rentals = rentalHistoryRepository.findByOwnerIdSnapshotAndEndDateBeforeAndStatusIn(
            ownerId,
            today,
            RentalOverduePolicy.statuses(),
            overduePageable(request.page(), request.size())
        )

        return toLentHistoryResponse(rentals, today)
    }

    // 현재 페이지의 등록자와 썸네일을 일괄 조회해 거래별 추가 조회가 발생하지 않게 합니다.
    private fun toBorrowedHistoryResponse(rentals: Page<Rental>, today: LocalDate): PageResponse<RentalHistoryResponse> {
        if (rentals.isEmpty) {
            return emptyResponse(rentals)
        }

        val equipmentIds = rentals.content.map(Rental::getEquipmentId).toSet()
        val ownerIds = rentals.content.map(Rental::getOwnerIdSnapshot).toSet()
        val userMap = userQueryPort.findSummaries(ownerIds)
        val thumbnailMap = loadThumbnails(equipmentIds)

        val responses = rentals.content.map { rental ->
            rentalHistoryMapper.toResponse(
                rental,
                getUser(userMap, rental.ownerIdSnapshot),
                thumbnailMap[rental.equipmentId],
                RentalOverduePolicy.calculateDays(rental, today)
            )
        }

        return toPageResponse(rentals, responses)
    }

    // 현재 페이지의 대여자와 썸네일을 일괄 조회해 거래별 추가 조회가 발생하지 않게 합니다.
    private fun toLentHistoryResponse(rentals: Page<Rental>, today: LocalDate): PageResponse<RentalHistoryResponse> {
        if (rentals.isEmpty) {
            return emptyResponse(rentals)
        }

        val renterIds = rentals.content.map(Rental::getRenterId).toSet()
        val equipmentIds = rentals.content.map(Rental::getEquipmentId).toSet()
        val userMap = userQueryPort.findSummaries(renterIds)
        val thumbnailMap = loadThumbnails(equipmentIds)

        val responses = rentals.content.map { rental ->
            rentalHistoryMapper.toResponse(
                rental,
                getUser(userMap, rental.renterId),
                thumbnailMap[rental.equipmentId],
                RentalOverduePolicy.calculateDays(rental, today)
            )
        }

        return toPageResponse(rentals, responses)
    }

    private fun loadThumbnails(equipmentIds: Collection<Long>): Map<Long, String> =
        equipmentThumbnailQueryPort.findThumbnailUrls(equipmentIds)

    private fun getUser(userMap: Map<Long, UserSummary>, userId: Long): UserSummary =
        userMap[userId] ?: throw CustomException(ErrorCode.USER_NOT_FOUND)

    private fun historyPageable(page: Int, size: Int): Pageable = PageRequest.of(
        page,
        size,
        Sort.by(Sort.Direction.DESC, "createdAt").and(Sort.by(Sort.Direction.DESC, "id"))
    )

    // 연체 이력은 가장 오래 지난 종료일을 먼저 보여주고 동일 종료일에서는 최신 거래를 우선합니다.
    private fun overduePageable(page: Int, size: Int): Pageable = PageRequest.of(
        page,
        size,
        Sort.by(Sort.Direction.ASC, "endDate")
            .and(Sort.by(Sort.Direction.DESC, "createdAt"))
            .and(Sort.by(Sort.Direction.DESC, "id"))
    )

    private fun normalizeKeyword(keyword: String?): String? = keyword?.trim()?.takeIf(String::isNotEmpty)

    private fun emptyResponse(rentals: Page<Rental>): PageResponse<RentalHistoryResponse> = PageResponse(
        emptyList(),
        rentals.number,
        rentals.size,
        rentals.totalElements,
        rentals.totalPages
    )

    private fun toPageResponse(rentals: Page<Rental>, responses: List<RentalHistoryResponse>): PageResponse<RentalHistoryResponse> = PageResponse(
        responses,
        rentals.number,
        rentals.size,
        rentals.totalElements,
        rentals.totalPages
    )
}
